server/src/main/java/org/elasticsearch/search/aggregations/support/CoreValuesSourceType.java

```diff
@@ -284,13 +283,13 @@ private ValuesSource.Numeric fieldData(FieldContext fieldContext, AggregationCon
             return new ValuesSource.Numeric.FieldData((IndexNumericFieldData) fieldContext.indexFieldData()) {
                 /**
                  * Proper dates get a real implementation of
-                 * {@link #roundingPreparer(IndexReader)}. If the field is
+                 * {@link #roundingPreparer(AggregationContext)}. If the field is
                  * configured with a script or a missing value then we'll
                  * wrap this without delegating so those fields will ignore
                  * this implementation. Which is correct.
                  */
                 @Override
-                public Function<Rounding, Rounding.Prepared> roundingPreparer() throws IOException {
+                public Function<Rounding, Rounding.Prepared> roundingPreparer(AggregationContext context) throws IOException {
```

Comment:
`AggregationContext` is only used here for extracting the `IndexReader`. Does it make sense to only pass the `IndexReader` here instead? 

Also, method `roundingPreparer` is only supported by  `DATE` fields. Does it make sense to keep this around in `ValuesSource` object?

Reply:
Just specifying IndexReader would suffice, so +1 to do that. But I already like that `AggregationContext` is less prominently visible in value source api. Maybe a lambda that return a min and max long would be even better? To less expose infrastructure classes in public methods? I'm ok with exploring this in a followup.

> Also, method roundingPreparer is only supported by DATE fields. Does it make sense to keep this around in ValuesSource object?

Maybe this date rounding should live in the places where it is used? (auto) data histogram aggs? I think this is the only place where it is used. Let's explore this in a follow up PR?

Reply:
I like the idea of passing in a lambda, this seems like a textbook use case for that pattern.  I'm also fine with doing that in a follow-up.

I _think_ the logic behind making this part of the general `ValuesSource` infrastructure instead of just something Date specific was that we might want to add different date-like fields in the future which might require different values sources.  I'm not sure that was great reasoning at the time, and it doesn't seem like we're moving towards doing anything like that right now, so I think limiting this method to Dates only would be fine.  We can always re-promote it if we need to.